### ISSUE TITLE
cloudfoundry-identity-client-lib should not depend on org.slf4j:slf4j-log4j12

### ISSUE THREAD

**Issue by kingd0m** [cloudfoundry/uaa#773]
Hello,

in my current project I get a warning by slf4j that multiple bindings are found.

By analyzing the reason I found that the cloudfoundry-identity-client-lib is using slf4j-log4j12 & slf4j-api. As of the SLF4J documentation libraries or frameworks should not declare a dependency on any SLF4J binding but only depend on slf4j-api (https://www.slf4j.org/codes.html#multiple_bindings)

If you agree with this, I would suggest to remove the slf4j-log4j12 dependency to be compliant with the slf4j recommendations.

Regards,
Michael

**Comment by DennisDenuto** [comment]
Hi @kingd0m, Thanks for creating this issue and investigating. Where did you see the warning about multiple slf4j bindings found?

**Comment by kingd0m** [comment]
Hi @DennisDenuto, I see it them when I start local the Tomcat Server. Prerequisite is, that I have defined an own binding in my application. And this than collides with the one which is additionally is defined by the  cloudfoundry-identity-client-lib.

Best Regards,
Michael

**Comment by DennisDenuto** [comment]
Hi @kingd0m 

I see, thanks for the explanation. Would something like this fix the warning you are seeing in your application: https://github.com/cloudfoundry/uaa/commit/d678c44a9d7e81a40db4d7fc6344fe6e0843f26c

**Comment by vtintillier** [comment]
Hi @DennisDenuto 

if it is no longer in the generated pom file, then yes. But ideally you should just never depend on a concrete slf4j implementation, only the -api one. It seems this is the culprit: https://github.com/cloudfoundry/uaa/blob/36e3fdfb520a9aef77ba8ad95542508ce2af6f3f/model/build.gradle#L18

I think this line should be removed.

Thanks

**Comment by DennisDenuto** [comment]
AFAIK In our application modules we do depend against `slf4j-api` (org.slf4j.Logger) when logging.

Our application depends on the binding`slf4j-log4j12` because that is the underlying logging framework we have decided to use. From the `slf4j` docs: " if no binding is found on the class path, then slf4j-api will default to a no-operation implementation discarding all log requests."

We do depend on `slf4j-log4j12` in some (a few config related classes) in our application code, i.e. org.apache.log4j.MDC (to instrument for distinguishing interleaved log
   output from different sources). Looks like it was a tradeoff between portability and using some specific log4j features.

Investigating this I did notice however we also depend on a binding for Jakarta Commons Logging (`jcl-over-sl4fj`). I am not sure if this was intentional or not.  I will ask the team and potentially track this in a separate issue.

I will make the change to the `client-lib` to stop depending on `log4j`.

**Comment by DennisDenuto** [comment]
@vtintillier good catch, looks like we are also depending on `slf4j-log4j12` in `cloudfoundry-identity-model`. 
We can also [remove](9c385357250972916b8e686a7ecac61853575e93) the `slf4j binding` from the model library too. 
I would be surprised if people are depending on the `model` library. If people are depending on  `client-lib` this [change](https://github.com/cloudfoundry/uaa/commit/e83e06107ed95c5df966d0604db8255e298d57d8) should prevent `slfj-log4j12` from coming in transitively.

**Comment by vtintillier** [comment]
Hello @DennisDenuto 

I tested your change locally. I added `apply plugin: 'maven-publish'` in the root `build.gradle` and ran `gradlew build publishToMavenLocal`. Then I looked in `$HOME/.m2/repository` and:
* `cloudfoundry-identity-model` pom file still depends on `slf4j-log4j12`, which is expected
* `cloudfoundry-identity-client-lib` pom file does not contain an exclusion of `slf4j-log4j12`, which is I guess what you tried to do with your change

**Comment by DennisDenuto** [comment]
Hi @vtintillier thanks for trying this out.

I just tried the above myself, but I got different results. for `model` and `client-lib` I don't see `slf4j-log4j12` listed as a dependency and is excluded from being transitively depended on too.
* i.e. for `cloudfoundry-identity-model` pom.xml
```
    <dependency>
      <groupId>junit</groupId>
      <artifactId>junit</artifactId>
      <version>4.12</version>
      <scope>test</scope>
      <exclusions>
        <exclusion>
          <artifactId>hamcrest-core</artifactId>
          <groupId>*</groupId>
        </exclusion>
        <exclusion>
          <artifactId>slf4j-log4j12</artifactId>
          <groupId>org.slf4j</groupId>
        </exclusion>
      </exclusions>
    </dependency>
```
* i.e. for `cloudfoundry-identity-client-lib` pom.xml
```
    <dependency>
      <groupId>org.cloudfoundry.identity</groupId>
      <artifactId>cloudfoundry-identity-model</artifactId>
      <version>4.12.0-SNAPSHOT</version>
      <scope>compile</scope>
      <exclusions>
        <exclusion>
          <artifactId>slf4j-log4j12</artifactId>
          <groupId>org.slf4j</groupId>
        </exclusion>
      </exclusions>
    </dependency>
```

This also matches what `./gradlew cloudfoundry-identity-model:dependencies` and `./gradlew cloudfoundry-identity-client-lib:dependencies` show.

**Comment by vtintillier** [comment]
Hello @DennisDenuto 

sorry for the late response, looks good then. Any chance it can get merged?

Thanks

**Comment by DennisDenuto** [comment]
fyi these changes have been pushed to develop
https://github.com/cloudfoundry/uaa/commit/adc526eea5d1927fb0e2087ed8133c8c7bd9743e
